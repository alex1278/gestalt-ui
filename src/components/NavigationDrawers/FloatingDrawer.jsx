import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const leftTransform = css`
  transform: translateX(${({ open, width }) => (open ? 0 : `-${width}`)});
  left: 0;
`;

const rightTransform = css`
  transform: translateX(${({ open, width }) => (open ? 0 : width)});
  right: 0;
`;

const baseAnime = css`
  transition: 225ms transform;
  transition-duration: ${({ open }) => (open ? '225ms' : '195ms')};
  transition-timing-function: cubic-bezier(0, 0, 0.2, 1);
  transition-delay: 0;
  will-change: transform;
`;

const Aside = styled.aside`
  position: fixed;
  top: 0;
  bottom: 0;
  flex: 1 0 auto;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0, 0, 0, 0.2), 0 6px 10px 0 rgba(0, 0, 0, 0.14), 0 1px 18px 0 rgba(0, 0, 0, 0.12);
  background-color: ${props => props.theme.colors.background};
  white-space: nowrap;
  width: ${props => props.width};
  ${baseAnime};
  ${props => (props.direction === 'left' ? leftTransform : rightTransform)};
`;

const FloatingDrawer = memo(({ open, width, direction, children }) => (
  <Aside open={open} direction={direction} width={width}>
    {children}
  </Aside>
));

FloatingDrawer.propTypes = {
  direction: PropTypes.oneOf(['left', 'right']),
  width: PropTypes.string,
  open: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

FloatingDrawer.defaultProps = {
  direction: 'left',
  width: '250px',
  open: false,
};

export default FloatingDrawer;
