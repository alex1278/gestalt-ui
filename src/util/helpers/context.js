/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
