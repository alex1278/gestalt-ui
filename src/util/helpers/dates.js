import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
