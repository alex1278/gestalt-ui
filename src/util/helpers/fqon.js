// sad hack for parsing url

export default function fqonParent(organization) {
  if (organization.org) {
    const urlParser = document.createElement('a');
    urlParser.href = organization.org.href;
    return urlParser.pathname.replace('/', '').split('/')[0];
  }

  return organization;
}
