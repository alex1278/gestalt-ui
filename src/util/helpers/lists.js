export function insertItem(array, payload) {
  return [...array, payload];
}

export function removeItem(array, payload) {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a === payload), 1);

  return newArray;
}

export function removeItemById(array, id, keyField = 'id') {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === id), 1);

  return newArray;
}

export function updateItem(array, payload) {
  return array.map((item) => {
    if (item.id === payload.id) {
      return {
        ...payload,
      };
    }

    // Leave every other item unchanged
    return item;
  });
}
