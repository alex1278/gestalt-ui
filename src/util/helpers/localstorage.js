/**
 * getItem
 * @param {*} key
 */
export const getItem = (key) => {
  try {
    return localStorage.getItem(key);
  } catch (err) {
    return undefined;
  }
};

/**
 * saveItem
 * @param {*} key
 * @param {*} value
 */
export const saveItem = (key, value) => {
  try {
    localStorage.setItem(key, value);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};

/**
 * removeItem
 * @param {*} key
 * @param {*} value
 */
export const removeItem = (key) => {
  try {
    localStorage.removeItem(key);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};
