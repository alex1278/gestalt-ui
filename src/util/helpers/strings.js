import { startCase, toLower } from 'lodash';

/**
 * getLastFromSplit
 * Parses and returns the last item in a structure like GrandPareng::Parent::Child
 * @param {String} string
 */
export function getLastFromSplit(string, delimiter = '::') {
  if (string) {
    const split = string.split(delimiter);

    return split[split.length - 1];
  }
  return string;
}

/**
 * truncate
 * @param {String} string
 * @param {Number} at
 * @param {String} trailing
 */
export function truncate(string, at, trailing = '...') {
  if (string) {
    if (string.length < at) {
      return string;
    }

    return `${string.substring(0, at).trim()}${trailing}`;
  }

  return string;
}

/**
 * toTitleCase
 * @param {String} string
 */
export function toTitleCase(string) {
  return startCase(toLower(string));
}


// TODO: Will be rmeoved when we have security config
export const checkIfPassword = field =>
  field && field.length && (field.toUpperCase().includes('PASSWORD') || field.toUpperCase().includes('SECRET') || field.toUpperCase().includes('KEY'));

export default {
  getLastFromSplit,
  truncate,
  toTitleCase,
  checkIfPassword
};

export const removeHostFromURL = (url) => {
  if (url) {
    return url.replace(/^.*\/\/[^/]+:?[0-9]?\//i, '');
  }

  return url;
};

export const isBase64 = (str) => {
  try {
    return btoa(atob(str)) === str;
  } catch (err) {
    return false;
  }
};
