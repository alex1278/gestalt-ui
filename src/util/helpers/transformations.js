import map from 'lodash/map';
import compact from 'lodash/compact';
import get from 'lodash/get';

/**
 * arrayToMap
 * @param {*} array
 * @param {*} keyName
 * @param {*} valueName
 */
export function arrayToMap(array = [], keyName = 'name', valueName = 'value') {
  if (!Array.isArray(array)) {
    return array;
  }

  return Object.assign({}, ...array.map(v => ({ [v[keyName]]: v[valueName] || '' })));
}

/**
 * mapTo2DArray
 * @param {*} object
 * @param {*} keyName
 * @param {*} valueName
 * @param {*} mergeSet
 */
export function mapTo2DArray(object = {}, keyName = 'name', valueName = 'value', mergeSet = {}) {
  if (typeof object !== 'object') {
    return object;
  }

  return map(object, (value, key) => (Object.assign({ [keyName]: key, [valueName]: value || '' }, mergeSet)));
}

/**
 * stringDemiltedToArray
 * convert comma delimited string to an array and remove blank entries
 * @param {*} string
 */
export function stringDemiltedToArray(string) {
  return compact(string.split(','));
}

/**
 * convertFromMaps
 * used to convert env vars from a map to an array and reconcile inherited vs own
 * @param {*} own
 * @param {*} inherited
 */
export function convertFromMaps(own = {}, inherited = {}, keyName = 'name', valueName = 'value') {
  const ownVars = mapTo2DArray(own, keyName, valueName, { inherited: false });

  // base case just return the mapped own vars
  if (!Object.keys(inherited)) {
    return ownVars;
  }

  const inheritedVars = mapTo2DArray(inherited, keyName, valueName, { inherited: true, overridden: false });

  const reconcileInheritedVars = inheritedVars.map((i) => {
    // find the dupe value in ownVars
    const index = ownVars.findIndex(o => o[keyName] === i[keyName]);

    // if the value is differnt mark the matching var as overriden
    if (index > -1 && ownVars[index][valueName] !== i[valueName]) {
      return { ...i, [valueName]: ownVars[index][valueName], inherited: true, overridden: true };
    }

    // default case the own env var
    return i;
  });

  const results = [...reconcileInheritedVars];

  // "merge" the ownVars back in
  ownVars.forEach((v) => {
    const index = inheritedVars.findIndex(o => o[keyName] === v[keyName]);

    if (index === -1) {
      results.push(v);
    }
  });

  return results;
}

/**
 * Builds an object from a delimited string
 * @param {String} string
 * @param {String} delimiter
 */
export function nestedObjectFromString(string, value = null, delimeter = '.') {
  const [last, ...paths] = string.split(delimeter).reverse();
  return paths.reduce((acc, el) => ({ [el]: acc }), { [last]: value });
}


export const getRoutePattern = (object, urlPattern) => {
  const keys = [];

  urlPattern.split('/').forEach((item) => {
    const search = item.startsWith(':');

    if (search) {
      keys.push(get(object, item.slice(1)));
    } else {
      keys.push(item);
    }
  });

  return keys.join('/');
};
