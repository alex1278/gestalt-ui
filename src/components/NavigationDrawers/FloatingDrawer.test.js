import React from 'react';
import FloatingDrawer from './FloatingDrawer';
import { renderWithTheme } from '../../../test/helpers';

describe('NaviationDrawers::FloatingDrawer', () => {
  it('renders correctly with basic props and closed', () => {
    const { container } = renderWithTheme(
      <FloatingDrawer title="Don't Mock me!">hahaha</FloatingDrawer>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with basic props when open', () => {
    const { container } = renderWithTheme(
      <FloatingDrawer title="Don't Mock me!" open>hahaha</FloatingDrawer>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when direction is right and closed', () => {
    const { container } = renderWithTheme(
      <FloatingDrawer title="Don't Mock me!" direction="right">hahaha</FloatingDrawer>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when direction is right and open', () => {
    const { container } = renderWithTheme(
      <FloatingDrawer title="Don't Mock me!" direction="right" open>hahaha</FloatingDrawer>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when a custom width is set', () => {
    const { container } = renderWithTheme(
      <FloatingDrawer title="Don't Mock me!" open width="100%">hahaha</FloatingDrawer>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
