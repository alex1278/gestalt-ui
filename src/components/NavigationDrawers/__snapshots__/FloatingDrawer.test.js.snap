// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NaviationDrawers::FloatingDrawer renders correctly when a custom width is set 1`] = `
.c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0,0,0,0.2),0 6px 10px 0 rgba(0,0,0,0.14),0 1px 18px 0 rgba(0,0,0,0.12);
  background-color: #fafafa;
  white-space: nowrap;
  width: 100%;
  -webkit-transition: 225ms -webkit-transform;
  -webkit-transition: 225ms transform;
  transition: 225ms transform;
  -webkit-transition-duration: 225ms;
  transition-duration: 225ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-delay: 0;
  transition-delay: 0;
  will-change: transform;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  left: 0;
}

<aside
  class="c0"
  direction="left"
  open=""
  width="100%"
>
  hahaha
</aside>
`;

exports[`NaviationDrawers::FloatingDrawer renders correctly when direction is right and closed 1`] = `
.c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0,0,0,0.2),0 6px 10px 0 rgba(0,0,0,0.14),0 1px 18px 0 rgba(0,0,0,0.12);
  background-color: #fafafa;
  white-space: nowrap;
  width: 250px;
  -webkit-transition: 225ms -webkit-transform;
  -webkit-transition: 225ms transform;
  transition: 225ms transform;
  -webkit-transition-duration: 195ms;
  transition-duration: 195ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-delay: 0;
  transition-delay: 0;
  will-change: transform;
  -webkit-transform: translateX(250px);
  -ms-transform: translateX(250px);
  transform: translateX(250px);
  right: 0;
}

<aside
  class="c0"
  direction="right"
  width="250px"
>
  hahaha
</aside>
`;

exports[`NaviationDrawers::FloatingDrawer renders correctly when direction is right and open 1`] = `
.c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0,0,0,0.2),0 6px 10px 0 rgba(0,0,0,0.14),0 1px 18px 0 rgba(0,0,0,0.12);
  background-color: #fafafa;
  white-space: nowrap;
  width: 250px;
  -webkit-transition: 225ms -webkit-transform;
  -webkit-transition: 225ms transform;
  transition: 225ms transform;
  -webkit-transition-duration: 225ms;
  transition-duration: 225ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-delay: 0;
  transition-delay: 0;
  will-change: transform;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  right: 0;
}

<aside
  class="c0"
  direction="right"
  open=""
  width="250px"
>
  hahaha
</aside>
`;

exports[`NaviationDrawers::FloatingDrawer renders correctly with basic props and closed 1`] = `
.c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0,0,0,0.2),0 6px 10px 0 rgba(0,0,0,0.14),0 1px 18px 0 rgba(0,0,0,0.12);
  background-color: #fafafa;
  white-space: nowrap;
  width: 250px;
  -webkit-transition: 225ms -webkit-transform;
  -webkit-transition: 225ms transform;
  transition: 225ms transform;
  -webkit-transition-duration: 195ms;
  transition-duration: 195ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-delay: 0;
  transition-delay: 0;
  will-change: transform;
  -webkit-transform: translateX(-250px);
  -ms-transform: translateX(-250px);
  transform: translateX(-250px);
  left: 0;
}

<aside
  class="c0"
  direction="left"
  width="250px"
>
  hahaha
</aside>
`;

exports[`NaviationDrawers::FloatingDrawer renders correctly with basic props when open 1`] = `
.c0 {
  position: fixed;
  top: 0;
  bottom: 0;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  height: 100%;
  z-index: 1000;
  box-shadow: 0 3px 5px -1px rgba(0,0,0,0.2),0 6px 10px 0 rgba(0,0,0,0.14),0 1px 18px 0 rgba(0,0,0,0.12);
  background-color: #fafafa;
  white-space: nowrap;
  width: 250px;
  -webkit-transition: 225ms -webkit-transform;
  -webkit-transition: 225ms transform;
  transition: 225ms transform;
  -webkit-transition-duration: 225ms;
  transition-duration: 225ms;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-delay: 0;
  transition-delay: 0;
  will-change: transform;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  left: 0;
}

<aside
  class="c0"
  direction="left"
  open=""
  width="250px"
>
  hahaha
</aside>
`;
